### PR TITLE
Cache character position in bytebuffer

### PR DESCRIPTION
**Pull request by klokane**
Provide better performance for mapping bytes to characters

### CODE REVIEW

**Comment by zdne** [comment]
@klokane where is this going to be used?

**Comment by klokane** [comment]
snowcrash, firstly make character map, later will use for generate SourceMaps instead of compute it on raw bytes.

**Comment by klokane** [comment]
There is optional solution, connect character index with, ByteBuffer together.

**Comment by klokane** [comment]
In this case we will need no make any changes to snowcrash

**Comment by zdne** [comment]
I am OK with this :+1:

File: src/ByteBuffer.cc [hunk 1]
```diff
@@ -52,6 +52,52 @@ static CharactersRange BytesRangeToCharactersRange(const BytesRange& bytesRange,
...
+void mdp::BuildCharacterIndex(ByteBufferCharacterIndex& index, const ByteBuffer& byteBuffer) {
+
+    const char* s = byteBuffer.c_str();
+    size_t len = byteBuffer.length();
+    size_t pos = 0;
+    size_t charPos = 0;
+
+    index.resize(byteBuffer.length());
+
+	while (s[pos] && pos < len) {
```
Comment (pksunkara) [inline src/ByteBuffer.cc]: Bad indentation for the code starting from this line. Please fix.

File: src/ByteBuffer.cc [hunk 2]
```diff
@@ -52,6 +52,52 @@ static CharactersRange BytesRangeToCharactersRange(const BytesRange& bytesRange,
...
+    if (bytesRange.length > 0)
+        charLength = index[bytesRange.location + bytesRange.length];
+    
+    CharactersRange characterRange = CharactersRange(charLocation, charLength);
+    return characterRange;
+}
+
+void mdp::BuildCharacterIndex(ByteBufferCharacterIndex& index, const ByteBuffer& byteBuffer) {
+
+    const char* s = byteBuffer.c_str();
```
Comment (pksunkara) [inline src/ByteBuffer.cc]: Please rename `s` to `source`.

**Pull request merged**